tests/test_prompt_creation.py

```diff
@@ -0,0 +1,140 @@
+            {"role": "user", "content": "is the car Song more expensive than car Tang?"},
+            {
+                "role": "assistant",
+                "content": None,
+                "function_call": {"name": "get_car_price", "arguments": '{\n  "car_name": "Song"\n}'},
+            },
+            {"role": "function", "content": "{'price': {'price': '$25000'}}", "name": "get_car_price"},
+            {
+                "role": "assistant",
+                "content": None,
+                "function_call": {"name": "get_car_price", "arguments": '{\n  "car_name": "Tang"\n}'},
+            },
+            {"role": "function", "content": "{'price': {'price': '$20000'}}", "name": "get_car_price"},
+            {
+                "role": "assistant",
+                "content": "No, the car Tang is less expensive than the car Song. The car Song is priced at $25,000, while the car Tang is priced at $20,000.",
+            },
+        ]
+        _, inputs = prepare_training_inputs(
+            test_case, tokenizer, padding="longest", max_length=256, return_tensor=False
+        )
+        input_ids = inputs["input_ids"]
+        labels = inputs["labels"]
+        self.assertEqual(len(input_ids), len(labels), "length of inputs and labels are different")
+
+        # check if input_ids[i] == labels[i] if labels[i] != -100
+        for input_token_id, label_token_id in zip(input_ids, labels):
+            if label_token_id != -100:
+                self.assertEqual(input_token_id, label_token_id, "input_token_id != label_token_id")
+
+        # Check if only messages where role=assistant are remained, others will be masked as -100
+        assistant_message = [item for item in test_case if item["role"] == "assistant"]
+        # find unmasked chunks in labels (chunk[i] != -100), there chunks are associated with assistant messages
+        # for example: labels=[-100, -100, 1, 2, 3, -100, -100, 4, 5] --> chunks = [[1,2,3], [4,5]]
+        chunks = extract_unmasked_chunks(labels)
+
+        self.assertEqual(
+            len(chunks),
+            len(assistant_message),
+            "number of unmasked chunks in labels is different from number of messages where role=assistant",
+        )
+        for chunk, message in zip(chunks, assistant_message):
+            decoded_content = "\nassistant" + tokenizer.decode(
+                chunk
+            )  # note that need to add: "\nassistant" because we mask this, see line 194 in prompt_utils.py
+            prompt = get_text_from_message(message)
+            # decoded_content and prompt should be the same
+            # to avoid any mistakes of tokenizer like adding white space we will compare after removing space
+            self.assertEqual(
+                re.sub("\s", "", decoded_content),
+                re.sub("\s", "", prompt),
+                "decoded content is different from original content",
+            )
+
```

Comment:
can we add a test that checks exact equality of an input and output of "get_prompt_from_messages"

Reply:
Sure